File: Java/Token.java
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.*;

public class Token {
    public enum TokenPrefix {
        SDK("NETLESSSDK_"),
        ROOM("NETLESSROOM_"),
        TASK("NETLESSTASK_");

        private String value;
        TokenPrefix(String name) {
            this.value = name;
        }
        public String getValue() {
            return value;
        }
    }

    /**
     * 数字越小，权限越大
     */
    public enum TokenRole {
        Admin("0"),
        Writer("1"),
        Reader("2");

        private String value;
        TokenRole(String name) {
            this.value = name;
        }
        public String getValue() {
            return value;
        }
    }

    public static String sdkToken(String accessKey, String secretAccessKey, long lifespan, Map<String, String> content) throws Exception {
        return createToken(TokenPrefix.SDK.getValue(), accessKey, secretAccessKey, lifespan, content);
    }

    public static String roomToken(String accessKey, String secretAccessKey, long lifespan, Map<String, String> content) throws Exception {
        return createToken(TokenPrefix.ROOM.getValue(), accessKey, secretAccessKey, lifespan, content);
    }

    public static String taskToken(String accessKey, String secretAccessKey, long lifespan, Map<String, String> content) throws Exception {
        return createToken(TokenPrefix.TASK.getValue(), accessKey, secretAccessKey, lifespan, content);
    }

    private static String createToken(String prefix, String accessKey, String secretAccessKey, long lifespan, Map<String, String> content) throws Exception {
        LinkedHashMap<String, String> map = new LinkedHashMap<>();
        map.putAll(content);
        map.put("ak", accessKey);
        map.put("nonce", UUID.randomUUID().toString());

        if (lifespan > 0) {
            map.put("expireAt", System.currentTimeMillis() + lifespan + "");
        }

        String information = toJson(sortMap(map));
        map.put("sig", createHmac(secretAccessKey, information));

        String query = sortAndStringifyMap(map);


        return prefix + stringToBase64(query);
    }

    private static LinkedHashMap<String, String> sortMap(Map<String, String> object) {
        List<String> keys = new ArrayList<>(object.keySet());
        keys.sort(null);

        LinkedHashMap<String, String> linkedHashMap = new LinkedHashMap<>();
        for (int i = 0; i < keys.size(); i++) {
            linkedHashMap.put(keys.get(i), object.get(keys.get(i)));
        }
        return linkedHashMap;
    }

    /**
     * 因程序本身的 map 只有一层，而非嵌套 map，所以此方法没有实现嵌套 map 转 string
     * 可自行替换为其他 json stringify 实现
     */
    private static String toJson(LinkedHashMap<String, String> map) {
        Iterator<Map.Entry<String, String>> iterator= map.entrySet().iterator();

        List<String> result = new ArrayList<>();
        while(iterator.hasNext()) {
            Map.Entry<String, String> entry = iterator.next();
            String value;
            if (entry.getValue() == null) {
                value = "null";
            } else {
                value = entry.getValue();
            }
            result.add("\"" + entry.getKey() + "\"" + ":" + "\"" + value + "\"");
        }
        return "{" + String.join(",", result) + "}";
    }

    private static String createHmac(String key, String data) throws Exception {
        Mac sha256_HMAC = Mac.getInstance("HmacSHA256");
        SecretKeySpec secret_key = new SecretKeySpec(key.getBytes("UTF-8"), "HmacSHA256");
        sha256_HMAC.init(secret_key);

        return byteArrayToHexString(sha256_HMAC.doFinal(data.getBytes("UTF-8")));
    }

    private static String sortAndStringifyMap(Map<String, String> object) {
        List<String> keys = new ArrayList<>(object.keySet());
        keys.sort(null);

        List<String> kvStrings = new ArrayList<>();
        for (int i = 0; i < keys.size(); i++) {
            if (object.get(keys.get(i)) == null) {
                continue;
            } else {
                kvStrings.add(encodeURIComponent(keys.get(i)) + "=" + encodeURIComponent(object.get(keys.get(i))));
            }
        }
        return String.join("&", kvStrings);
    }

    private static String stringToBase64(String str) throws UnsupportedEncodingException {
        return Base64.getEncoder().encodeToString(str.getBytes("utf-8")).replace("+", "-").replace("/", "_").replaceAll("=+$", "");
    }

    private static String byteArrayToHexString(byte[] b) {
        StringBuilder hs = new StringBuilder();
        String stmp;
        for (int n = 0; b!=null && n < b.length; n++) {
            stmp = Integer.toHexString(b[n] & 0XFF);
            if (stmp.length() == 1)
                hs.append('0');
            hs.append(stmp);
        }
        return hs.toString().toLowerCase();
    }

    /**
     * encodeURIComponent 基于 url 编码对字符串进行编码
     * 最终实现和 JavaScript 中的 encodeURIComponent 一致
     *
     * https://stackoverflow.com/questions/607176/java-equivalent-to-javascripts-encodeuricomponent-that-produces-identical-outpu
     */
    private static String encodeURIComponent(String s) {
        String result = null;

        try {
            result = URLEncoder.encode(s, "UTF-8")
                    .replaceAll("\\+", "%20")
                    .replaceAll("%21", "!")
                    .replaceAll("%27", "'")
                    .replaceAll("%28", "(")
                    .replaceAll("%29", ")")
                    .replaceAll("%7E", "~");
        }
        // This exception should never occur.
        catch (UnsupportedEncodingException e) {
            result = s;
        }

        return result;
    }
}

File: golang/token.go
package token

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"github.com/google/uuid"
	"net/url"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

// 数字越小，权限越大
const (
	AdminRole  = "0"
	WriterRole = "1"
	ReaderRole = "2"
)

const (
	sdkPrefix  = "NETLESSSDK_"
	roomPrefix = "NETLESSROOM_"
	taskPrefix = "NETLESSTASK_"
)

type SDKContent struct {
	Role string
}

type RoomContent struct {
	Role string
	Uuid string
}

type TaskContent struct {
	Role string
	Uuid string
}

// SDKToken 生成 sdk token
func SDKToken(accessKey string, secretAccessKey string, lifespan int64, content *SDKContent) string {
	m := map[string]string{
		"role": content.Role,
	}
	return createToken(sdkPrefix)(accessKey, secretAccessKey, lifespan, &m)
}

// RoomToken 生成 room token
func RoomToken(accessKey string, secretAccessKey string, lifespan int64, content *RoomContent) string {
	m := map[string]string{
		"role": content.Role,
		"uuid": content.Uuid,
	}
	return createToken(roomPrefix)(accessKey, secretAccessKey, lifespan, &m)
}

// TaskToken 生成 task token
func TaskToken(accessKey string, secretAccessKey string, lifespan int64, content *TaskContent) string {
	m := map[string]string{
		"role": content.Role,
		"uuid": content.Uuid,
	}
	return createToken(taskPrefix)(accessKey, secretAccessKey, lifespan, &m)
}

// bufferToBase64 buffer 转 base64
// 并格式化字符
func bufferToBase64(b []byte) string {
	str := base64.StdEncoding.EncodeToString(b)

	// 替换 "+" 到 "-"
	// 替换 "/" 到 "_"
	{
		r := strings.NewReplacer(
			"+", "-",
			"/", "_",
		)
		str = r.Replace(str)
	}

	// 移除末尾所以的 "="
	// 例如: hello== -> hello
	{
		r := regexp.MustCompile("=+$")
		str = r.ReplaceAllString(str, "")
	}

	return str
}

// getMapKeys 提取 map 里的 key
// 并且以 key 为主进行排序
func getMapKeys(m *map[string]string) []string {
	keys := make([]string, len(*m))

	i := 0
	for k := range *m {
		keys[i] = k
		i++
	}

	sort.Strings(keys)
	return keys
}

// encodeURIComponent 基于 url 编码对字符串进行编码
// 最终实现和 JavaScript 中的 encodeURIComponent 一致
func encodeURIComponent(str string) string {
	r := url.QueryEscape(str)

	// golang 的 url.QueryEscape 实现和 JavaScript 中的 encodeURIComponent 不一致的地方是对空格处理
	// golang 里会把空格转成 "+"，所以这里通过字符串替换，再把 "+" 转成 "%20"，以保证和 js 的实现一致
	r = strings.ReplaceAll(r, "+", "%20")
	return r
}

// stringify 序列化 Map
// 实现逻辑可参考: https://github.com/sindresorhus/query-string/blob/master/index.js#L284
func stringify(m *map[string]string) string {
	keys := getMapKeys(m)

	var arr []string
	for _, k := range keys {
		if (*m)[k] != "" {
			arr = append(arr, encodeURIComponent(k)+"="+encodeURIComponent((*m)[k]))
		}
	}

	return strings.Join(arr, "&")
}

// mergeMap 合并两个 map，返回新的 map 对象
// 在有相同值的情况下，m2 的优先级更高
func mergeMap(m1, m2 *map[string]string) map[string]string {
	result := make(map[string]string)

	for k, v := range *m1 {
		result[k] = v
	}

	for k, v := range *m2 {
		result[k] = v
	}

	return result
}

// jsonStringify 模范 JavaScript 中 JSON.stringify 的实现
// 因程序本身的 map 只有一层，而非嵌套 map，所以此方法没有实现嵌套 map 转 string
func jsonStringify(m *map[string]string) string {
	keys := getMapKeys(m)

	var body []string
	for _, k := range keys {
		body = append(body, "\""+k+"\""+":"+"\""+(*m)[k]+"\"")
	}

	result := strings.Join(body, ",")
	return "{" + result + "}"
}

// expireAt 根据当时时间及用户传入毫秒数，来计算过期时间
func expireAt(lifespan int64) string {
	return strconv.FormatInt(time.Now().UTC().UnixNano()/1e6+lifespan, 10)
}

// hmac256 实现 hmac 加密并转成 hex，实现可参考 nodejs: crypto/createHmac 及 digest("hex")
func hmac256(data string, secret string) string {
	key := []byte(secret)
	h := hmac.New(sha256.New, key)

	h.Write([]byte(data))

	return hex.EncodeToString(h.Sum(nil))
}

// createToken 根据 prefix 生成相应的 token
func createToken(prefix string) func(string, string, int64, *map[string]string) string {
	return func(accessKey string, secretAccessKey string, lifespan int64, content *map[string]string) string {
		m := map[string]string{
			"ak":    accessKey,
			"nonce": uuid.Must(uuid.NewRandom()).String(),
		}
		m = mergeMap(content, &m)

		if lifespan > 0 {
			m["expireAt"] = expireAt(lifespan)
		}

		m["sig"] = hmac256(jsonStringify(&m), secretAccessKey)

		query := stringify(&m)
		return prefix + bufferToBase64([]byte(query))
	}
}
